### ISSUE TITLE
Calibre Says That DRM Is Still Present

### ISSUE THREAD

**Issue by HinkleMcKrinkle** [ReverseApple/RemoveBooksDRM#10]
First off, thank you for taking this on for us! I cloned V2 and ran the build.sh but it threw a ton of warnings. I'll attach the output to this issue. I am not versed in this so if it's something I'm just missing, any help would be awesome. The strange thing is that I am still able to run the inject.sh script and it goes through the entire process to the "Should've opened" prompt and spits out a decrypted epub but when I try to import that epub into calibre, it still sees a DRM. 

- M1 macbook pro | macOS Sequoia 15.1.1
- Books 7.1 (6269)
- Calibre 7.22.0
[output.txt](https://github.com/user-attachments/files/18204037/output.txt)

**Comment by AngeloD2022** [comment]
Hi, can you confirm that `antidote.dylib` is in the same directory as `inject.sh`?

**Comment by HinkleMcKrinkle** [comment]
Yup, it is in the same directory. attached a screen shot of the file structure. 
![Screenshot 2024-12-20 at 4 32 06 AM](https://github.com/user-attachments/assets/81a7704e-0ab4-4890-a8d3-0cf1ce4cc1e1)

**Comment by AngeloD2022** [comment]
> Yup, it is in the same directory. attached a screen shot of the file structure. 
> 
> ![Screenshot 2024-12-20 at 4 32 06 AM](https://github.com/user-attachments/assets/81a7704e-0ab4-4890-a8d3-0cf1ce4cc1e1)
> 
> 

What does it say when you run:

```
file path/to/antidote.dylib
```

**Comment by HinkleMcKrinkle** [comment]
file ~/.RemoveBooksDRM/antidote.dylib
/Users/XXX/.RemoveBooksDRM/antidote.dylib: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64e:Mach-O 64-bit dynamically linked shared library arm64e]

/Users/XXX/.RemoveBooksDRM/antidote.dylib (for architecture x86_64):	Mach-O 64-bit dynamically linked shared library x86_64

/Users/XXX/.RemoveBooksDRM/antidote.dylib (for architecture arm64e):	Mach-O 64-bit dynamically linked shared library arm64e

**Comment by AngeloD2022** [comment]
Oh boy. This is **super** dumb... I'm really not sure how this mistake managed to escape me and made its way to the release.

This line inside of `inject.sh` is likely the problem:
```
SCRIPT_DIR=$(cd "$(dirname "$0")"; pwd);
```

Try replacing it with this:
```
SCRIPT_DIR=$(pwd)
```

**Comment by AngeloD2022** [comment]
hmm wait... no that's correct I think.

**Comment by HinkleMcKrinkle** [comment]
I deleted my original repo and cloned the repo again, but when I ran the build.sh it threw the same errors I had in my original comment. I also corrected the inject.sh but the epub it spit out still has a DRM in calibre. 

Build.sh output
[output2.txt](https://github.com/user-attachments/files/18213799/output2.txt)

**Comment by HinkleMcKrinkle** [comment]
This error pops up twice when I run the inject.sh as well. I looked it up and it seemed like a non-issue but i'll throw it in here just in case. 

 Books[9475:148405] Unbalanced calls to begin/end appearance transitions for <BKUpgradeViewController: 0x13d876450>.

**Comment by AngeloD2022** [comment]
Would you mind sharing the name of the asset you're trying to unprotect?

**Comment by HinkleMcKrinkle** [comment]
Flying Through Midnight ISBN13: 9780743281997 
Raise the Titanic! ISBN: 9780425194522

**Comment by aosterhage** [comment]
I've just attempted to use v2.0 myself (universal zip, rather than build from scratch) and ran into the same issue: it appears that the DRM is still intact to Calibre. I also attempted to open the decrypted archive in the Books app and it says that the archive is formatted incorrectly:

> Cannot Open “How to Think Like a Roman Emperor”
> It is formatted incorrectly, or is not a format that Apple Books can open.

I do expect this second error after some conversion though, I seem to recall that the Apple Books app utilizes directories instead of archives for EPUB formats.

The terminal output from the inject.sh script is attached, it appears to me like everything "worked" successfully.
[log.txt](https://github.com/user-attachments/files/18270558/log.txt)

The only other information I can think to add is that this particular book is an EPUB within my Books directory. I was under the impression that this could then be converted and DRM removed via this script. If there's a different indication as to if this book is actually iBook formatted, please correct me.

**Comment by AngeloD2022** [comment]
@aosterhage Interesting...

If the book was in an iBooks format, it should present an entirely different dialog (see below).

https://github.com/ReverseApple/RemoveBooksDRM/blob/ac41f7451b04cd88a3f8230e8dafcc162399169d/src/main.m#L180

I believe the tool is prompting you with decryption options, right? If you extract the resulting EPUB file, could you confirm that the interior asset files are decrypted?

**Comment by aosterhage** [comment]
Okay, I took some time to actually inspect the output from the decryption. The contents *are* decrypted. If I unzip the EPUB archive and point Calibre to the `package.opf` manually, it does open the book perfectly.

There were 3 'extra' files in the `META-INF/` directory that I think might be causing the problem: `encryption.xml`, `signatures.xml`, and `sinf.xml`. These appear to be left over from the DRM protected EPUB. I removed these 3 files, created the archive again, and now everything seems to work perfectly. Calibre detects no DRM and reads the book great. I'm guessing that Calibre has a simple check to see if the EPUB is DRM protected: if `encryption.xml` or other files exist, then it 'must' be DRM protected.

I haven't taken a look at the code for this tool yet, but the fix should be as simple as ensuring that the DRM files are removed from the archive.

@tduncan907 I'm going to try to explain the steps I took so you can do the same if you would like. They require using the command line (Terminal app) so if you aren't comfortable with that, it might be best to wait for the fix.

1. Unzip the .epub into a separate folder.
    1. `mkdir <some empty directory>`, e.g. `mkdir ~/Downloads/my-ebook/`
    2. `unzip <epub archive> <some empty directory>`, e.g. `unzip MyEbook.epub ~Downloads/my-ebook/`
2. Remove the 3 files I discussed above.
    3. `cd ~/Downloads/my-ebook/META-INF/`
    4. `rm encryption.xml signatures.xml sinf.xml`
3. Zip up the ebook directory back to a .epub archive
    1. `cd ~/Downloads/my-ebook` (move back to the root of the epub directory)
    2. `zip -X <new epub archive> mimetype`, e.g. `zip -X MyNewEbook.epub mimetype`
    3. `zip -rg MyNewEbook.epub META-INF -x \*.DS_Store`
    4. `zip -rg MyNewEbook.epub OEBPS -x \*.DS_Store`

**Comment by AngeloD2022** [comment]
> There were 3 'extra' files in the META-INF/ directory that I think might be causing the problem: encryption.xml, signatures.xml, and sinf.xml.

Yep. Forgot to remove those... I'll need to make a change that does it automatically.

**Comment by HinkleMcKrinkle** [comment]
Cloned the updated repo and it worked perfectly! Thanks to both of you for the help. Wish I could've been as helpful as  @aosterhage to troubleshoot with you @AngeloD2022
